iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp

```diff
@@ -88,30 +65,49 @@ struct TorchIndexSelectOpConversion
     mhlo::TorchIndexSelectOp::Adaptor adaptor(args);
     int axis = op.dim();
     int batch = op.batch_dims();
-    auto indexShapeType = adaptor.index().getType().dyn_cast<ShapedType>();
+    auto indexShapeType = adaptor.index().getType().cast<ShapedType>();
     int nIndices = indexShapeType.getRank();
-    auto inputShapeType = adaptor.input().getType().dyn_cast<ShapedType>();
+    auto inputShapeType = adaptor.input().getType().cast<ShapedType>();
     if (axis < 0) axis += inputShapeType.getRank();
     if (batch < 0) batch += nIndices;
 
     Location loc = op.getLoc();
-    Value output = op.getResult();
-    int rank = output.getType().cast<ShapedType>().getRank();
+    ShapedType resultType = op.getResult().getType().cast<ShapedType>();
+    int rank = resultType.getRank();
+
     SmallVector<AffineMap, 2> indexingMaps;
     SmallVector<AffineExpr, 4> exprs;
-    for (int i = 0; i < batch; ++i)
+    for (int i = 0; i < batch; ++i) {
       exprs.push_back(rewriter.getAffineDimExpr(i));
-    for (int i = 0, e = nIndices - batch; i < e; ++i)
+    }
+    for (int i = 0, e = nIndices - batch; i < e; ++i) {
       exprs.push_back(rewriter.getAffineDimExpr(axis + i));
+    }
     indexingMaps.emplace_back(
         AffineMap::get(rank, /*symbolCount=*/0, exprs, rewriter.getContext()));
     indexingMaps.emplace_back(rewriter.getMultiDimIdentityMap(rank));
+
     SmallVector<StringRef, 3> loopTypes(rank, getParallelIteratorTypeName());
-    ShapedType outputType = op.getResult().getType().cast<ShapedType>();
-    Value initOp = getOutputTensor(rewriter, loc, op.getResult());
-    if (!initOp) return failure();
+
+    // The output shape is
+    //   `params[:axis] + indices[batch_dims:] + params[axis + 1:]`
+    SmallVector<Value, 4> dynSizes;
+    for (int i = 0; i < rank; ++i) {
+      if (!resultType.isDynamicDim(i)) continue;
+      if (i < axis) {
+        dynSizes.push_back(rewriter.create<DimOp>(loc, adaptor.input(), i));
```

Comment:
Nit: Use createOrFold here.

Reply:
It only benefits static shapes because it will be folded into a constant. Since it is not easy to support mixed shapes, let's keep it this way. (see explanation below)